Refactored argument passing in backup and restore to use a context variable

jmcatamney opened the pull request:
Previously, the functions and helper classes in gpcrondump, gpdbrestore, dump.py,
and restore.py all required long lists of arguments to be passed around and
required many helper functions.  This made adding new command-line options
or new features quite time-consuming to implement and to test.

Now, a Context class has been added to backup_utils.py that holds all of the
variables used for command-line options, as well as other variables used in
many places during backup and restore; additionally, many helper functions
have been condensed into a few functions within Context, and in the future
more such functions can be moved to Context as needed.  This will make it
much easier to expand dump and restore functionality going forward.

jimmyyih commented:
+10 on the concept

Maintaining the code should be a lot easier with this commit.

larham commented on gpMgmt/bin/gpdbrestore:
```diff
@@ -49,18 +41,23 @@ logger = gplog.get_default_logger()
 
 class GpdbRestore(Operation):
     def __init__(self, options, args):
+        context = Context(options)
```
In other places, the reference to Context was saved as self.context = Context(options).  Should that be done here?

jmcatamney replied:
This context gets assigned to self.context at the end of __init__, so it works either way, but I'll change it to match gpcrondump for consistency.

larham commented on gpMgmt/bin/gpdbrestore:
```diff
@@ -169,207 +133,167 @@ class GpdbRestore(Operation):
         os._exit(0)
 
     def _output_info(self, info, restore_type):
+        context = self.context
```
Why create a local variable here, context = self.context?  The next developer might think that the object was ephemeral, as opposed to seeing self.context and knowing that the context object is long-lived.

jmcatamney replied:
That's a leftover from before I standardized everything to use self.context directly.  I'll change that here and wherever else it might still be.

larham commented on gpMgmt/bin/gpdbrestore:
```diff
@@ -169,207 +133,167 @@ class GpdbRestore(Operation):
 
-        if len(self.restore_tables) > 0:
-            logger.info("Database name              = %s" % info['restore_db'])
+        if context.restore_tables is not None:
```
This test diff is different... old code is checking for len > 0, while new code just wants non-null.  Is that correct?

jmcatamney replied:
Looks like a merge issue after the special characters story.  I'll change that to use the len() check.

larham commented on gpMgmt/bin/gppylib/test/behave/mgmt_utils/steps/mgmt_utils.py:
```diff
@@ -1498,10 +1499,12 @@ def impl(context, table, dbname, ao_table):
     ao_sch, ao_tbl = ao_table.split('.') 
     part_info = [(1, ao_sch, ao_tbl, tbl)]
     try:
+        backup_utils = Context()
```
in other places, this kind of local var is called "context".  Is there a good reason to call it "backup_utils"?

kaknikhil replied:
maybe call it backup_utils_context ??

larham commented on gpMgmt/bin/gppylib/test/behave/mgmt_utils/steps/netbackup_mgmt_utils.py:
```diff
@@ -191,86 +189,46 @@ def impl(context, options):
 @when('verify that {filetype} file has been backed up using netbackup')
 @then('verify that {filetype} file has been backed up using netbackup')
 def impl(context, filetype):
+    backup_utils = Context()
```
consider renaming "backup_utils" to "context"

jmcatamney replied:
Behave already uses a variable named "context" for all of its steps, as can be seen in the line above the change; "backup_utils" is a compromise to avoid a collision.

larham commented on gpMgmt/bin/gpdbrestore:
```diff
@@ -169,207 +133,167 @@ class GpdbRestore(Operation):
-            cmdline = 'gpddboost --sync --dir=%s' % self.master_datadir
-            if self.ddboost_storage_unit:
-                cmdline += ' --ddboost-storage-unit %s' % self.ddboost_storage_unit
+        context = self.context
```
Why create a local variable for context?

jmcatamney replied:
Fixed as per the previous comment on local context variables.

larham commented on gpMgmt/bin/gpdbrestore:
```diff
@@ -450,186 +353,144 @@ class GpdbRestore(Operation):
+            temp = match[len(prefix):]
+            self.context.timestamp = temp[0:14]
 
         # TODO: do a local GetDbName after Scp
```
Is this TODO removable?  see line 464,   self.context.restore_db = GetDbName(filename).run()

jmcatamney replied:
Looks like it is.  Will do.

larham commented on gpMgmt/bin/gpdbrestore:
```diff
@@ -69,97 +66,64 @@ class GpdbRestore(Operation):
 
-        if self.change_schema and len(self.schema_level_restore_list) > 0:
-            raise Exception('-S schema level restore does not work with --change-schema option')
+        if context.table_file is not None:
```
consider shortening to simply 
if context.table_file:

also line 105, etc.

jmcatamney replied:
I've now removed all the "is not None"s and changed the "foo is None"s to "not foo"s.

larham commented on gpMgmt/bin/gpdbrestore:
```diff
@@ -450,186 +353,144 @@ class GpdbRestore(Operation):
-            (restore_timestamp, restore_db, compress) = self._validate_db_date_dir()
-        elif self.db_host_path is not None:
-            (restore_timestamp, restore_db, compress, dump_host, dump_file) = self._validate_db_host_path()
+        if self.context.timestamp is not None:
```
consider shortening to simply
if self.context.timestamp:

also line 362, 363, etc.

kaknikhil commented:
gpcrondump : line 63 - why do we only care about a subset of attributes from options. like "clear_dumps" . what about cleanup_date , cleanup_total etc. ?

jmcatamney commented:
@kaknikhil Those are redundant, only the non-context variables should be assigned there.  I'll remove those.

kaknikhil commented:
gpcrondump :
  * line 341 - old code "len(self.dump_databases) == 0) " vs new code "self.context.dump_databases is None)" ,not the same check right ?

jmcatamney commented:
The check should be "len(self.context.dump_databases) = 0", since dump_databases is guaranteed to be non-None, I'll fix that.

kaknikhil commented:
gpcrondump

  * 1087 - no need to pass the context object. It's possible that there are more methods like _prompt_continue in gpcrondump which don't need to be passed a context object. Can you please check ?

jmcatamney commented:
It's been fixed, and I didn't see any other instances of that.

kaknikhil commented:
gpcrondump
  * 948 in the old code : did we intend to remove the code or was it moved somewhere else ?

kaknikhil commented on gpMgmt/bin/gppylib/operations/backup_utils.py:
```diff
@@ -9,12 +9,140 @@
+        "dump": ("dump_%d_%d", ""), "files": ("dump", "_regular_files"), "filter": ("dump", "_filter"), "global": ("global_1_1", ""),
+        "increments": ("dump", "_increments"), "last_operation": ("dump", "_last_operation"), "master_config": ("master_config_files", ".tar"),
+        "metadata": ("dump_1_1", ""), "partition_list": ("dump", "_table_list"), "pipes": ("dump", "_pipes"), "plan": ("restore", "_plan"),
+        "postdata": ("dump_1_1", "_post_data"), "report": ("dump", ".rpt"), "schema": ("dump", "_schema"), "segment_config": ("segment_config_files_%d_%d", ".tar"),
```
maybe a newline after postdata's value so that "report" is in it's own separate line.

kaknikhil commented:
backup_utils.py
  * 609 - backup_file_with_ddboost only needs 2 arguments i.e context and filetype , right ?
  * 620 - no need to have timestamp as an argument

kaknikhil commented on gpMgmt/bin/gpdbrestore:
```diff
@@ -49,18 +41,23 @@ logger = gplog.get_default_logger()
 class GpdbRestore(Operation):
     def __init__(self, options, args):
+        context = Context(options)
+        context.options_list = " ".join(sys.argv[1:])
```
can we do this inside the Context class so that we don't have to repeat this for both dump and restore?

jmcatamney replied:
Since options_list isn't passed down from gpcrondump or gpdbrestore, I'll just take it out of context and leave it as self.options_list.

kaknikhil commented on gpMgmt/bin/gpdbrestore:
```diff
@@ -49,18 +41,23 @@ logger = gplog.get_default_logger()
+        context.options_list = " ".join(sys.argv[1:])
+        self.search_for_dbname = options.search_for_dbname
+        if self.search_for_dbname and '"' in self.search_for_dbname:
+            self.search_for_dbname = self.search_for_dbname.strip('"')
```
do we really want to strip all the double quotes ?? this code will strip all the double quotes at the start and end of the string. what if the dname has double quotes in the start or end like ""foo"" ?? we will end up with the string foo which would be incorrect .

jmcatamney replied:
This is a leftover from the special characters merge and will be removed.

kaknikhil commented on gpMgmt/bin/gpdbrestore:
```diff
@@ -169,207 +133,167 @@ class GpdbRestore(Operation):
+            host, path = context.db_host_path
+            RecoverRemoteDumps(context, host = host, path = path).run()
 
-        if len(self.restore_tables) > 0 and self.truncate:
```
was this code moved somewhere else ?

jmcatamney replied:
It's now in restore.py under RestoreDatabase, to simplify the redirected restore logic.

jmcatamney commented:
gpcrondump

70: Yes, this is redundant with the assignment in context and will be removed.

1023: Both pipe_list and file_list are used, as they're written to their respective files, but dump_dir is no longer needed and will be removed.

1239: The calls to verify_lines_in_file were removed deliberately, as gpcrondump was essentially writing to a file and then immediately reading it back, which is unnecessary since if we can't trust the filesystem we have bigger problems.  The function has been left in because other utilities (primarily analyzedb) imports from backup_utils.py so it was better to leave it in a central location than to reimplement it in several other places.

1175: Already fixed as per the previous comment on redundant self.context passing.

backup_utils

609: backup_file_with_ddboost allows you to either pass context and filetype (to have it generate a filename and transfer that file) or context and path (to transfer the specified file).  If filetype is passed, dbid and timestamp may be necessary for generating some files.  The hostname parameter is unnecessary in either case, though, and will be removed.

620: The timestamp argument is necessary for constructing the --to-file parameter if a timestamp was used to create the file passed to the function.

kaknikhil commented on gpMgmt/bin/gppylib/operations/backup_utils.py:
```diff
@@ -137,37 +263,11 @@ def check_successful_dump(report_file_contents):
-        dirname = get_ddboost_backup_directory()
-        dirname = "%s/%s" % (dirname, timestamp[0:8])
-    return "%s/%sgp_cdatabase_1_1_%s" % (dirname, dump_prefix, timestamp)
+    return context.generate_filename("cdatabase", timestamp=timestamp)
```
how does generate_filename take care of ddboost , don't we need to rely on the show-config code to get the ddboost backup directory ??

kaknikhil commented on gpMgmt/bin/gppylib/operations/backup_utils.py:
```diff
@@ -9,12 +9,140 @@
+            filename += ".gz"
+        return filename
+
+    def generate_prefix(self, filetype, dbid=1, timestamp=None):
```
are we making sure that ddboost is taken care of while getting the backup directory ?

jmcatamney replied:
The existing logic for get_backup_dir/get_backup_directory doesn't account for ddboost, it assumes that backup_dir is already set to the appropriate location.

kaknikhil commented on gpMgmt/bin/gppylib/operations/buildMirrorSegments.py:
```diff
@@ -633,6 +633,21 @@ def createConfigureNewSegmentCommand(hostName, cmdLabel, validationOnly):
+            for destSeg in destSegments:
+                if srcSeg.content == destSeg.content:
+                    cmd = Scp('copy db_dumps from old segment to new segment',
+                               srcSeg.getSegmentDataDirectory() + '/db_dumps/*',
```
consider using os.path.join ...

jmcatamney replied:
Will do.

kaknikhil commented:
dump.py :- 
  * 494 we are not using the timestamp value from lines 486-492, should timestamp be assigned to context.timestamp? If not , we can remove the code lines in 486-492. 
  * 532 config_backup_file is not being used

chrishajas commented on gpMgmt/bin/gpdbrestore:
```diff
@@ -169,207 +133,167 @@ class GpdbRestore(Operation):
                 logger.info("Drop and re-create db      = Off")
 
-        if self.db_timestamp is not None and len(self.restore_tables) == 0:
+        if context.timestamp is not None and context.restore_tables is None:
```
This (and a few other checks) is checking if restore_tables is None, while other checks on restore_tables test if len > 0 (line 288 for example). I see that restore_tables is initialized to None. Should these all be is None checks?

jmcatamney replied:
These have already been fixed as per a previous comment on "is not None"s.  As a side note, backup_utils now initializes restore_tables to an empty array given that the CLI default is an empty array.

chrishajas commented:
In dump.py:

line 1075: Is the include_dump_tables_file variable ever None? It is initialized to an empty string in backup_utils.py
1106: same thing

In backup_utils.py, these are initialized to empty strings, which aren’t None.  There might be a few more of these scattered about.

jmcatamney commented:
dump.py
494: The timestamp should have been used in the generate_filename call, I've corrected that.

532: I've folded that into the backup_file_with_nbu call.

1075 and 1106: I'll change the "foo is not None"s to "not foo"s.

kaknikhil commented:
restore.py 

  * 63 & 65 - not being used , we are re declaring both the variables again
  * 560 table_file is re declared later 
  * 613 file is not used anywhere
  * 944 - 957 mdd gets assigned to self.context.master_datadir and in line 957 self.context.master_datadir gets assigned to mdd without any changes to mdd in between. looks like some leftover code ?

jmcatamney commented:
restore.py

63, 65, 560, 613: Fixed.

944-957: It's not mdd that gets changed, mdd is saving the value of master_datadir as that is changed to each segment data directory in turn.

1211: I've refactored the function slightly to avoid the issue.

Chibin commented on gpMgmt/bin/gppylib/test/behave/mgmt_utils/steps/netbackup_mgmt_utils.py:
```diff
@@ -191,86 +189,46 @@ def impl(context, options):
-        cmd.run(validateAfter=True)
-        if cmd.get_results().stdout.strip() != filename:
-            raise Exception('Cdatabase file %s was not backup up to NetBackup server %s successfully' % (filename, netbackup_service_host))
+        for type in ['ao', 'co', 'last_operation']:
```
So we're getting rid of the 'cdatabase' check?

jmcatamney replied:
It wasn't removed; the check on line 202 now handles everything except config and state files, which includes the cdatabase file.

Chibin replied:
**not** in ['config', 'state']: 
got it

jmcatamney commented:
Pushed to master: https://github.com/greenplum-db/gpdb/commit/9d1ae66cd92f0eab7111541b677c0d2de11e5acb
